File: solutions/python/grade-school/1/grade_school.py
from collections import defaultdict


class School:
    _students: dict[int, set[str]]
    _add_log: list[bool]

    def __init__(self) -> None:
        self._students = defaultdict(set)
        self._add_log = []

    def add_student(self, name: str, grade: int) -> None:
        if any(n == name for ns in self._students.values() for n in ns):
            self._add_log.append(False)
            return
        self._students[grade].add(name)
        self._add_log.append(True)

    def roster(self) -> list[str]:
        return [
            name
            for _, names in sorted(self._students.items())
            for name in sorted(names)
        ]

    def grade(self, grade: int) -> list[str]:
        return sorted(self._students[grade])

    def added(self) -> list[bool]:
        return self._add_log

File: solutions/c/grade-school/1/grade_school.c
#include "grade_school.h"

#include <string.h>

void init_roster(roster_t* roster) { roster->count = 0; }

bool add_student(roster_t* roster, const char* name, uint8_t grade) {
    if (roster->count >= MAX_STUDENTS) {
        return false;
    }
    for (size_t i = 0; i < roster->count; ++i) {
        if (strcmp(roster->students[i].name, name) == 0) {
            return false;
        }
    }
    for (size_t i = 0; i < roster->count; ++i) {
        if (roster->students[i].grade > grade ||
            (roster->students[i].grade == grade &&
             strcmp(roster->students[i].name, name) > 0)) {
            for (size_t j = roster->count; j > i; --j) {
                roster->students[j] = roster->students[j - 1];
            }
            roster->students[i].grade = grade;
            strcpy(roster->students[i].name, name);
            roster->count++;
            return true;
        }
    }
    student_t* student = &roster->students[roster->count++];
    student->grade = grade;
    strcpy(student->name, name);
    return true;
}

roster_t get_grade(const roster_t* roster, uint8_t grade) {
    roster_t result;
    init_roster(&result);
    for (size_t i = 0; i < roster->count; ++i) {
        if (roster->students[i].grade == grade) {
            add_student(&result, roster->students[i].name, grade);
        }
    }
    return result;
}
